### ISSUE TITLE
Configuration issues with AVS driver release

### ISSUE THREAD

**Issue by xinato** [thesofproject/avs-topology-xml#5]
It's great that the support for the older platforms is going forward with AVS! I have some confusion about configuration. I have the nami board which uses the snd-soc-kbl_da7219_max98357a driver.  I am using Fedora 37.

1. Do I have to disable the old skl stack to use AVS? Is blacklisting the snd-soc-skl module sufficient? If I blacklist it, I also have to inhibit it from loading as a dependency.

2. The nami board has four PCMs, HDA through three HDMI ports, DMIC input, max98357a through the speakers, and da7219 for the headset. Do I need to configure all four topology files to make this work? Do I have to load all four AVS board files?

3. How should the topology files be named so that the driver finds them properly? I am currently digging through the source files to try to answer this one.

4. Are the topology files in the git useful as they are, or will they need more configuration to match the hardware?

I appreciate all the great work that is going into this development. I am eager to get it going.

Best regards,

Chris Murphy

**Comment by xinato** [comment]
I did get the avsdk working so that I can convert the xml files to ucs and then to tplg bin files. I am still uncertain about the file names. I still don't see any alsa hw: devices.

**Comment by crojewsk-intel** [comment]
The 'skylake-driver -> avs-skylake' transition is about to commence for the end-users. If it was up to developers alone, the transition would have already happened : ) Unfortunately, that's not the case. I appreciate the patience.

1. There is a module specifically for that called `snd_intel_dspcfg`, its code lies in sound/hda in the Linux kernel. The module parameter you are looking for is called `dsp_driver`, available options are listed in [intel-dsp-config.h](https://elixir.bootlin.com/linux/latest/source/include/sound/intel-dsp-config.h#L13). Essentially you can impact the selection by adding/modifying:
```
options snd_intel_dspcfg dsp_driver=4 # for the avs-driver
#options snd_intel_dspcfg dsp_driver=2 # for the skylake-driver
```

2. The avs-driver embraces granular card approach - there is one board/topology pair per actual device. Nami has four of those, HDMI, DMIC (internal mic), max98357a (speaker) and da7219 (headphones). To answer your question: yes, four pairs are required to have a complete setup. As the design is granular, missing any of these will still allow AudioDSP to boot along with any devices that have appropriate topology file available to them e.g.: lack of HDMI topology won't prevent DMIC from enumerating.

3. The expected names are listed in the [board_selection.c](https://elixir.bootlin.com/linux/latest/source/sound/soc/intel/avs/board_selection.c) file that is part of the Linux kernel.

4. All files present here are useful. What's mostly missing are bits related to 3rd party modules which I cannot really comment on. Good example are amplifiers e.g.: maxXXXXX codecs where usage of external module safeguards the hardware from frying. Here, these can be replaced by Peakvol module which is part of the base firmware and which we can help you configure without any limitations : )

**Comment by crojewsk-intel** [comment]
What will happen soon is addition of new audio formats to files available here in the avsdk project so that all most commonly used ones like: 16b/16b and 24b/32b are part of the topology by default. Given how many emails we received regarding maxXXXXX amplifiers, Peakvol module is going to be added by default to Speaker path too.

**Comment by xinato** [comment]
Thank you for your kind and detailed response. I have tried to follow it.

I updated my modprobe.conf file with:

options snd_intel_dspcfg dsp_driver=4

I compiled (xml -> conf -> bin) files into the firmware directory:

$ ls /lib/firmware/intel/avs
da7219-tplg.bin  dmic-tplg.bin  max98357a-tplg.bin

I still get:

$ cat /proc/asound/cards
--- no soundcards ---

I see the following in my log:

Feb 14 14:08:22 bluey kernel: snd_soc_avs 0000:00:1f.3: bound 0000:00:02.0 (ops i915_audio_component_bind_ops [i915])
Feb 14 14:08:22 bluey kernel: snd_soc_avs 0000:00:1f.3: Direct firmware load for intel/avs/skl/dsp_basefw.bin failed with error -2
Feb 14 14:08:22 bluey kernel: snd_soc_avs 0000:00:1f.3: request firmware failed: -2
Feb 14 14:08:22 bluey kernel: snd_soc_avs 0000:00:1f.3: firmware boot failed: -2


But I cannot find the file dsp_basefw.bin anywhere in the firmware directory.


These modules:
snd_soc_avs_da7219
snd_soc_avs_dmic
snd_soc_avs_max98357a

Do not load at boot time. I can manually load them using modprobe, but that does not result in an operating sound system.

I have attached the output of lsmod here. Note that the above three modules where loaded manually.

[lsmod.log](https://github.com/thesofproject/avs-topology-xml/files/10736198/lsmod.log)



I suspect my first issue is the missing dsp_basefw.bin file. I am unsure where to get this.

Best regards, and many thanks.
chris

**Comment by xinato** [comment]
After poking around a bit, it looks like I need firmware revision 9.21.0.4732 or newer. The latest revision in the Fedora distribution is 9.21.0.3420. I suspect that this new version will be hard to get hold of. That will be disappointing since the skl driver quit working for me after kernel 6.1. I am stuck at kernel 6.0.18. I understand that most of the work is going into the avs driver. I am not expecting to see much more work on the skl driver at this point.

**Comment by WeirdTreeThing** [comment]
You can add `options snd-soc-avs ignore_fw_version=1` to your modprobe config to get around the firmware revision requirement.

**Comment by crojewsk-intel** [comment]
Updated binaries have been sent to the linux-firmware mailing list: [\[PULL v2\] intel: Update AudioDSP firmware for BDW/SKL/APL/CNL platforms](https://lore.kernel.org/linux-firmware/38913aba-3717-fa26-bac8-96f8e9cf112a@intel.com/T/#u).

I'd avoid using any kind of outdated AudioDSP firmware binary. There is a reason the `min_fw_version` is in place. Module parameter `ignore_fw_version` is mainly for debug purposes as engineering firmware versions may not have a valid version number.

**Comment by crojewsk-intel** [comment]
@xinato to make transition as painless as possible we still support users with any bugs they encounter. I bet you speak of [recent skylake-driver problems on Eve and other Chromebooks](https://lore.kernel.org/regressions/CALFERdzKUodLsm6=Ub3g2+PxpNpPtPq3bGBLbff=eZr9_S=YVA@mail.gmail.com). Fix has been provided to upstream with [\[PATCH\] ALSA: hda: Fix codec device field initialization](https://lore.kernel.org/alsa-devel/20230210165541.3543604-1-cezary.rojewski@intel.com/).

**Comment by xinato** [comment]
Mr. Crojewsk, you are full of surprises! Thank you very much for taking interest in my issues and providing so much help. Even though I have the problems below, I think I have great progress thanks to your help.

I installed the DSP firmware and got over that hurdle. The directory you directed me to had another file, dsp_mod_7CAD0808-AB10-CD23-EF45-12AB34CD56EF.bin. I put that file in the same directory, /lib/firmware/intel/avs/skl. I did not see anything that referred to this file in the logs.

I got the HDA->HDMI and the bluetooth connections working with sound. That is the great news.

The DA7219 connection shows up in in the device list (aplay -l or pavucontrol) but I don't hear anything on the headphones. Capture does not seem to work either through the headset.

$ aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: hdaudioB0D2 [hdaudioB0D2], device 1: HDMI1 (*) []
  Subdevices: 0/1
  Subdevice #0: subdevice #0
card 0: hdaudioB0D2 [hdaudioB0D2], device 2: HDMI2 (*) []
  Subdevices: 0/1
  Subdevice #0: subdevice #0
card 0: hdaudioB0D2 [hdaudioB0D2], device 3: HDMI3 (*) []
  Subdevices: 0/1
  Subdevice #0: subdevice #0
card 1: avsda7219 [avs_da7219], device 1: Headset (*) []
  Subdevices: 0/1
  Subdevice #0: subdevice #0

The max98357a device did not seem to enumerate. The snd_soc_max98357a driver loads, but the snd_soc_avs_max98357 driver does not.

On the capture side, the DMIC does not show up in the device list. The headset shows up in the list, but it captures only silence.

$ arecord -l
**** List of CAPTURE Hardware Devices ****
card 1: avsda7219 [avs_da7219], device 1: Headset (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0

The DMIC has the following message in the log:

Feb 15 14:51:17 bluey kernel: avs_dmic avs_dmic: ASoC: no complete control IO handler for DMIC Volume type (g,p,i) 257:257:1 at 0x1990
Feb 15 14:51:17 bluey kernel: avs_dmic avs_dmic: ASoC: failed to load widget DMIC FE
Feb 15 14:51:17 bluey kernel: avs_dmic avs_dmic: ASoC: topology: could not load header: -22
Feb 15 14:51:17 bluey kernel: snd_soc_avs 0000:00:1f.3: load topology "intel/avs/dmic-tplg.bin" failed: -22
Feb 15 14:51:17 bluey kernel: snd_soc_avs 0000:00:1f.3: ASoC: error at snd_soc_component_probe on dmic-platform: -22
Feb 15 14:51:17 bluey kernel: avs_dmic avs_dmic: ASoC: failed to instantiate card -22
Feb 15 14:51:17 bluey kernel: avs_dmic: probe of avs_dmic failed with error -22

So I still have to troubleshoot the snd_soc_avs_max98357a issue and the proper operation of the da7219 (render and capture).

My log also has a kernel oops that is similar to one I saw in the skl driver. The oops is attached. In both cases it happens where setting up gtw_cfg for the dsp.

Again, many thanks for your help. If you want any more information, please ask.

Best regards,
Chris
[oops.log](https://github.com/thesofproject/avs-topology-xml/files/10748050/oops.log)

**Comment by xinato** [comment]
Added lsmod list and files in firmware directories

[lsmod&files.log](https://github.com/thesofproject/avs-topology-xml/files/10751129/lsmod.files.log)

**Comment by xinato** [comment]
So more good news. I found the problem with sound out of the da7219 was caused by a mixer setting. I expect the capture will be the same when I try it in the morning.

**Comment by xinato** [comment]
The problem with the headset capture does seem to be the mixer. Looks like I have to develop use case files for each device.

**Comment by crojewsk-intel** [comment]
> Mr. Crojewsk, you are full of surprises! Thank you very much for taking interest in my issues and providing so much help. Even though I have the problems below, I think I have great progress thanks to your help.

You're welcome.

> I installed the DSP firmware and got over that hurdle. The directory you directed me to had another file, dsp_mod_7CAD0808-AB10-CD23-EF45-12AB34CD56EF.bin. I put that file in the same directory, /lib/firmware/intel/avs/skl. I did not see anything that referred to this file in the logs.

The module in question is part of the debugging suite. While for APL and onward it is part of the base AudioDSP firmware, for firmware branch targeting KBL-based devices it is not the case. We share it just in case to speed up the debug should the module usage be necessary. Here, we are bringing up software, so no need for firmware debug : )

> The max98357a device did not seem to enumerate. The snd_soc_max98357a driver loads, but the snd_soc_avs_max98357 driver does not.

This is probably caused by a missing entry in board_selection.c for max98357a. Addressed in v6.2 with [\[PATCH\] ASoC: Intel: avs: Add missing audio amplifier for KBL](https://lore.kernel.org/all/Y3wHyJ%2FEcsLRHGr3@tora/).

> On the capture side, the DMIC does not show up in the device list. The headset shows up in the list, but it captures only silence.
> 
> $ arecord -l **** List of CAPTURE Hardware Devices **** card 1: avsda7219 [avs_da7219], device 1: Headset (*) [] Subdevices: 1/1 Subdevice #0: subdevice #0
> 
> The DMIC has the following message in the log:
> 
> Feb 15 14:51:17 bluey kernel: avs_dmic avs_dmic: ASoC: no complete control IO handler for DMIC Volume type (g,p,i) 257:257:1 at 0x1990 Feb 15 14:51:17 bluey kernel: avs_dmic avs_dmic: ASoC: failed to load widget DMIC FE Feb 15 14:51:17 bluey kernel: avs_dmic avs_dmic: ASoC: topology: could not load header: -22 Feb 15 14:51:17 bluey kernel: snd_soc_avs 0000:00:1f.3: load topology "intel/avs/dmic-tplg.bin" failed: -22 Feb 15 14:51:17 bluey kernel: snd_soc_avs 0000:00:1f.3: ASoC: error at snd_soc_component_probe on dmic-platform: -22 Feb 15 14:51:17 bluey kernel: avs_dmic avs_dmic: ASoC: failed to instantiate card -22 Feb 15 14:51:17 bluey kernel: avs_dmic: probe of avs_dmic failed with error -22

Similarly to above. This time it's lack of Peakvolume support, added with [\[PATCH 0/4\] Add support for DSP volume controls](https://lore.kernel.org/all/20221214185500.3896902-1-amadeuszx.slawinski@linux.intel.com/).

> My log also has a kernel oops that is similar to one I saw in the skl driver. The oops is attached. In both cases it happens where setting up gtw_cfg for the dsp.

> Best regards, Chris [oops.log](https://github.com/thesofproject/avs-topology-xml/files/10748050/oops.log)

Thanks for the report. This is a known issue - not an oops, just an info. It's a false positive though. A byproduct of "better safe than sorry" changes in code that manages memory-related operations. You can view recent history of files such as `include/linux/fortify-string.h` to see them for yourself, it's quite a range of patches. I agree with those changes, stacks such as ones present in your log may obscure the dmesg a bit and generate some false positives, but those checks are useful and make the kernel more secure in general.

**Comment by crojewsk-intel** [comment]
> The problem with the headset capture does seem to be the mixer. Looks like I have to develop use case files for each device.

We certainly did not prepare any UseCaseManagement (UCM) files i.e.: files that list the controls which have to be set/switched to have given endpoint enabled. Will be addressed in time - obviously telling users to do that manually with `amixer` and such is not convenient.

**Comment by xinato** [comment]
Thank you for the update, Mr. Crojewski. It looks like kernel 6.2 will be landing shortly. Hopefully, that will solve my issues with DMIC and max98357a. 

In the short term, I will work on the UCM for the da7219 card. I have some experience with the UCM for the skl driver. But the skl was a single card. Since there are four different cards associated with the avs driver, I will need to familiarize myself with how to make transitions between cards. For example, when the headset plug is inserted, the devices have to change and the routing has to change from one card to another. The UCM files are associated with a single card.

I also have to work out  a safe maximum volume level for the mx98357a to protect the speakers. If I read the patch correctly, it sounds like this may have to be configured in the topology file.

Thanks again.
Best Regards,
Chris

**Comment by xinato** [comment]
I have had some time to work on this again finally.

I downloaded kernel-6.2.1-200.fc38.x86_64 and the topology file from https://github.com/thesofproject/avs-topology-xml/tree/for-skylake-driver. This combination of kernel 6.2.1, skl driver, and topology file gave me a very usable system. Speakers, headset, DMIC, and HDMI all work properly.

-----------------

Switching to the AVS driver is still problematic. The avs_DMIC driver enumerates, but cannot load the topology. I installed the DMIC topology file in the /lib/firmware/intel/avs directory. I get the following error:

Mar 07 17:23:31 bluey kernel: avs_dmic avs_dmic: ASoC: no complete control IO handler for DMIC Volume type (g,p,i) 257:257:1 at 0x1990
Mar 07 17:23:31 bluey kernel: avs_dmic avs_dmic: ASoC: failed to load widget DMIC FE
Mar 07 17:23:31 bluey kernel: avs_dmic avs_dmic: ASoC: topology: could not load header: -22
Mar 07 17:23:31 bluey kernel: snd_soc_avs 0000:00:1f.3: load topology "intel/avs/dmic-tplg.bin" failed: -22
Mar 07 17:23:31 bluey kernel: snd_soc_avs 0000:00:1f.3: ASoC: error at snd_soc_component_probe on dmic-platform: -22
Mar 07 17:23:31 bluey kernel: avs_dmic avs_dmic: ASoC: failed to instantiate card -22

If I get some time, I will compile a version of the AVS driver with some additional debug logging to figure out why it cannot load the topology.

With the avs driver, I can play audio through the ALSA devices. For some reason, pipewire is still failing. I haven't spent a lot of time figuring out the pipewire problem because I need the dmic to work for a usable system first. So I am continuing to use the skl driver.

Also, with AVS installed and pipewire disabled, the kernel panics every time I shut down. I will attach a photo of the panic screen. I hope this is useful information. I am willing to do more testing if you have something to suggest.

**Comment by xinato** [comment]
This is the screenshot of the kernel panic that occurs when I shut down while the avs driver is running and pipewire is disabled using the fedora kernel-6.2.1-200.fc38.x86_64.

![PXL_20230308_014646957](https://user-images.githubusercontent.com/80137266/223600070-0a682a52-d5bf-46e5-8d47-00c73b164231.jpg)

**Comment by xinato** [comment]
I also tested the combination of kernel-6.1.14-200.fc37.x86_64 with the skl driver and the topology file from https://github.com/thesofproject/avs-topology-xml/tree/for-skylake-driver. This combination also appears to work well.

**Comment by crojewsk-intel** [comment]
Thanks for your feedback @xinato. You even spent time taking the picture! Resolution is fine, we were able to get everything we needed. <3

Now back to the technical stuff. The avs-driver is not officially the default yet - this have to be kept in mind. While the patches are in Mark's (our ASoC subsystem maintainer) tree, it does not mean they are immediately propagated to linux-stable. It takes time. Propagating driver-supporting files also takes time - right now topology files are only published here and are not installed in /lib/firmware by default whenever you update or upgrade your distro.

In regard to:
```
Mar 07 17:23:31 bluey kernel: avs_dmic avs_dmic: ASoC: no complete control IO handler for DMIC Volume type (g,p,i) 257:257:1 at 0x1990
Mar 07 17:23:31 bluey kernel: avs_dmic avs_dmic: ASoC: failed to load widget DMIC FE
Mar 07 17:23:31 bluey kernel: avs_dmic avs_dmic: ASoC: topology: could not load header: -22
Mar 07 17:23:31 bluey kernel: snd_soc_avs 0000:00:1f.3: load topology "intel/avs/dmic-tplg.bin" failed: -22
Mar 07 17:23:31 bluey kernel: snd_soc_avs 0000:00:1f.3: ASoC: error at snd_soc_component_probe on dmic-platform: -22
Mar 07 17:23:31 bluey kernel: avs_dmic avs_dmic: ASoC: failed to instantiate card -22
```
it's clear that the kernel of yours is missing: [\[PATCH 0/4\] Add support for DSP volume controls](https://lore.kernel.org/all/20221214185500.3896902-1-amadeuszx.slawinski@linux.intel.com/). These are not part of v6.1, they were merged later.

In regard to kernel panic on shutdown - this revealed a flaw in our integration infrastructure. You see, regardless of how many machines we have connected there and kernel version iterated, tests utilize only "pure" ALSA utilities e.g.: aplay/arecord and such. `rm pulseaudio` in /usr/bin and similar makes quick work of any upper layer so the CI is sure nothing intervenes with it. This has its pros - isolation of kernel driver testing, no userspace noise. Cons include problems such as one you posted here. There were few reports so one of our developers took the task and addressed it with: [\[PATCH 1/3\] ASoC: Intel: avs: Implement PCI shutdown](https://lore.kernel.org/alsa-devel/20230113190310.1451693-2-amadeuszx.slawinski@linux.intel.com).

Testing the skylake-driver with topology published here at for-skylake-driver branch did not go unnoticed - Thank you. When publishing avs-driver stuff we decided to keep downstream in mind. While costly, should return the favor, shouldn't it? And thus the skylake-driver supporting files are here too. Looks like we did the right thing : )

**Comment by xinato** [comment]
Thank you for the update, Cezary. I will be patient waiting for the release of the driver and support files. I hope that I am not the only one who appreciates the support you are giving to the skl driver. Thank you for getting me back to a working system for now. Yes, you did the right thing for me.

**Comment by xinato** [comment]
My system continues to work with the skl driver as described earlier (thank you very much!). With the release of Fedora's kernel 6.3.1, I tried the AVS driver again. aplay and arecord show the following setup:

$ aplay -l
**** List of PLAYBACK Hardware Devices ****
card 2: avsda7219 [avs_da7219], device 1: Headset (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 3: hdaudioB0D2 [hdaudioB0D2], device 1: HDMI1 (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 3: hdaudioB0D2 [hdaudioB0D2], device 2: HDMI2 (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 3: hdaudioB0D2 [hdaudioB0D2], device 3: HDMI3 (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 4: avsmax98357a [avs_max98357a], device 1: Built-in Speakers (*) []
  Subdevices: 0/1
  Subdevice #0: subdevice #0

$ arecord -l
**** List of CAPTURE Hardware Devices ****
card 1: avsdmic [avs_dmic], device 2: Digital Microphone (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 2: avsda7219 [avs_da7219], device 1: Headset (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
[cjm@bluey sound]$ 


This looks good, but I was not able to play sound using  aplay -Dhw:4,0 tst_c2_f16.wav. I enabled debugging in  snd-intel-sdw-acpi, snd-hda-core, and snd-intel-dspcfg. This results in the following errors:

May 07 09:14:47 bluey kernel: snd_intel_dspcfg:intel_nhlt_get_endpoint_blob:298: snd_soc_avs 0000:00:1f.3: Looking for configuration:
May 07 09:14:47 bluey kernel: snd_intel_dspcfg:intel_nhlt_get_endpoint_blob:299: snd_soc_avs 0000:00:1f.3:   vbus_id=5 link_type=3 dir=0, dev_type=4
May 07 09:14:47 bluey kernel: snd_intel_dspcfg:intel_nhlt_get_endpoint_blob:301: snd_soc_avs 0000:00:1f.3:   ch=2 fmt=24/32 rate=48000
May 07 09:14:47 bluey kernel: snd_intel_dspcfg:intel_nhlt_get_endpoint_blob:302: snd_soc_avs 0000:00:1f.3: Endpoint count=4
May 07 09:14:47 bluey kernel: snd_intel_dspcfg:nhlt_check_ep_match:271: snd_soc_avs 0000:00:1f.3: Endpoint: vbus_id=0 link_type=2 dir=1 dev_type = 1
May 07 09:14:47 bluey kernel: snd_intel_dspcfg:nhlt_check_ep_match:271: snd_soc_avs 0000:00:1f.3: Endpoint: vbus_id=1 link_type=3 dir=0 dev_type = 4
May 07 09:14:47 bluey kernel: snd_intel_dspcfg:nhlt_check_ep_match:271: snd_soc_avs 0000:00:1f.3: Endpoint: vbus_id=1 link_type=3 dir=1 dev_type = 4
May 07 09:14:47 bluey kernel: snd_intel_dspcfg:nhlt_check_ep_match:271: snd_soc_avs 0000:00:1f.3: Endpoint: vbus_id=0 link_type=3 dir=0 dev_type = 4
May 07 09:14:47 bluey kernel: snd_soc_avs 0000:00:1f.3: no I2S ep_blob found
May 07 09:14:47 bluey kernel: snd_soc_avs 0000:00:1f.3: module-type create failed: -2
May 07 09:14:47 bluey kernel: snd_soc_avs 0000:00:1f.3: error creating module -2
May 07 09:14:47 bluey kernel: snd_soc_avs 0000:00:1f.3: create path failed: -2
May 07 09:14:47 bluey kernel: snd_soc_avs 0000:00:1f.3: ASoC: error at snd_soc_dai_hw_params on SSP0 Pin: -2
May 07 09:14:47 bluey kernel:  SSP0-Codec: ASoC: error at __soc_pcm_hw_params on SSP0-Codec: -2
May 07 09:14:47 bluey kernel:  Built-in Speakers: ASoC: error at dpcm_fe_dai_hw_params on Built-in Speakers: -2
May 07 09:14:47 bluey pipewire[2363]: pw.link: 0x558e84412780: one of the nodes is in error out:suspended in:error
May 07 09:14:47 bluey pipewire[2363]: pw.link: 0x558e84414270: one of the nodes is in error out:suspended in:error

I expect that the sound would play on the "Built-in Speakers". It looks like the driver is searching for an ACPI entry matching the platfrom audio controller instead of the avsmax98357a. This match fails and prevents the sound from playing.

I hope this is helpful.

**Comment by crojewsk-intel** [comment]
> I hope this is helpful.

Without feedback such as this, we're in the dark as the amount of configurations in our labs is limited, primarily to commercial configurations.

As I do care about the growth of the community, you can improve the readability of your posts by differentiating logs from the body of your message - usage of code tags (```) is sufficient. An example below.

Now, some time ago I've conducted the process of making Intel's [NHLT specification](https://01.org/sites/default/files/595976_intel_sst_nhlt.pdf) public. It has also been adopted by iASL [some time ago](https://bugs.acpica.org/show_bug.cgi?id=1525) too. Mentioning this as one can help themselves by simply viewing what their hardware/BIOS is actually prepared to support. The doc/tool along with avsdk provide that functionality among their other capabilities.

The log reads:
```
May 07 09:14:47 bluey kernel: snd_intel_dspcfg:intel_nhlt_get_endpoint_blob:298: snd_soc_avs 0000:00:1f.3: Looking for configuration:
May 07 09:14:47 bluey kernel: snd_intel_dspcfg:intel_nhlt_get_endpoint_blob:299: snd_soc_avs 0000:00:1f.3: vbus_id=5 link_type=3 dir=0, dev_type=4
May 07 09:14:47 bluey kernel: snd_intel_dspcfg:intel_nhlt_get_endpoint_blob:301: snd_soc_avs 0000:00:1f.3: ch=2 fmt=24/32 rate=48000
May 07 09:14:47 bluey kernel: snd_intel_dspcfg:intel_nhlt_get_endpoint_blob:302: snd_soc_avs 0000:00:1f.3: Endpoint count=4
May 07 09:14:47 bluey kernel: snd_intel_dspcfg:nhlt_check_ep_match:271: snd_soc_avs 0000:00:1f.3: Endpoint: vbus_id=0 link_type=2 dir=1 dev_type = 1
May 07 09:14:47 bluey kernel: snd_intel_dspcfg:nhlt_check_ep_match:271: snd_soc_avs 0000:00:1f.3: Endpoint: vbus_id=1 link_type=3 dir=0 dev_type = 4
May 07 09:14:47 bluey kernel: snd_intel_dspcfg:nhlt_check_ep_match:271: snd_soc_avs 0000:00:1f.3: Endpoint: vbus_id=1 link_type=3 dir=1 dev_type = 4
May 07 09:14:47 bluey kernel: snd_intel_dspcfg:nhlt_check_ep_match:271: snd_soc_avs 0000:00:1f.3: Endpoint: vbus_id=0 link_type=3 dir=0 dev_type = 4
```
Looking for endpoint described with:
- Virtual bus ID = 5, also known as port number e.g.: SSP port number
- Link type = 3, means SSP link
- Device type = 4, means SSP Analog Codec
- Direction=0, means playback

So, driver is trying to find BLOB for codec device connected to SSP5 and given the link type, it will be utilizing offloaded path, I2S transfer facilitated by our AudioDSP firmware. Audio format it's looking for: 24/32/2/48000.

NHLT present on your machine supports the following 4 endpoints:
- PDM (DMIC), capture, port=0, in case of DMIC it means queue but that's a separate subject, irrelevant here
- SSP Analog codec connected to port 1, blob for playback stream
- SSP Analog codec connected to port 1, blob for capture stream
- SSP Analog codec connected to port 0, blob for playback stream

Judging by the available BLOBs, one can tell the device supports internal mic (DMIC), headphones (pb/cp both on port1) and speakers (just pb on port0). Command used: `aplay -Dhw:4,1 (...)` and your `aplay -l` tells us that speaker device was requested for playback. In this case it's max98357a. Information available in [board_selection.c](https://elixir.bootlin.com/linux/latest/source/sound/soc/intel/avs/board_selection.c) tells us that said codec is supported by the avs-driver on SSP0 or SSP5 depending on the platform type. Virtual bus ID=5 narrows the list to: APL-based device as that's the only one on the list with SSP5.

**TLDR: Playback attempted on APL-based device using NHLT for non-APL-based device - SKL/KBL.**

Guess simply updating the Virtual bus ID would do the trick as the rest of the BLOB does not care about it.

**Comment by WeirdTreeThing** [comment]
I'm having the same issue with max98357a on my skylake chromebook (cave).
> TLDR: Playback attempted on APL-based device using NHLT for non-APL-based device - SKL/KBL.

Seems like the opposite issue, playback on SKL/KBL device using the APL SSP port instead of the correct one for the platform.

In my dmesg I have:
```
[ 1342.854587] snd_intel_dspcfg:intel_nhlt_get_endpoint_blob: snd_soc_avs 0000:00:1f.3: Looking for configuration:
[ 1342.854592] snd_intel_dspcfg:intel_nhlt_get_endpoint_blob: snd_soc_avs 0000:00:1f.3:   vbus_id=5 link_type=3 dir=0, dev_type=4
[ 1342.854598] snd_intel_dspcfg:intel_nhlt_get_endpoint_blob: snd_soc_avs 0000:00:1f.3:   ch=2 fmt=24/32 rate=48000
[ 1342.854604] snd_intel_dspcfg:intel_nhlt_get_endpoint_blob: snd_soc_avs 0000:00:1f.3: Endpoint count=4
[ 1342.854608] snd_intel_dspcfg:nhlt_check_ep_match: snd_soc_avs 0000:00:1f.3: Endpoint: vbus_id=0 link_type=2 dir=1 dev_type = 1
[ 1342.854613] snd_intel_dspcfg:nhlt_check_ep_match: snd_soc_avs 0000:00:1f.3: Endpoint: vbus_id=0 link_type=3 dir=0 dev_type = 4
[ 1342.854619] snd_intel_dspcfg:nhlt_check_ep_match: snd_soc_avs 0000:00:1f.3: Endpoint: vbus_id=1 link_type=3 dir=0 dev_type = 4
[ 1342.854624] snd_intel_dspcfg:nhlt_check_ep_match: snd_soc_avs 0000:00:1f.3: Endpoint: vbus_id=1 link_type=3 dir=1 dev_type = 4
[ 1342.854642] snd_soc_avs 0000:00:1f.3: no I2S ep_blob found
[ 1342.854644] snd_soc_avs 0000:00:1f.3: module-type create failed: -2
[ 1342.854647] snd_soc_avs 0000:00:1f.3: error creating module -2
[ 1342.854758] snd_soc_avs 0000:00:1f.3: create path failed: -2
[ 1342.854762] snd_soc_avs 0000:00:1f.3: ASoC: error at snd_soc_dai_hw_params on SSP0 Pin: -2
[ 1342.854767]  SSP0-Codec: ASoC: error at __soc_pcm_hw_params on SSP0-Codec: -2
[ 1342.854772]  Built-in Speakers: ASoC: error at dpcm_fe_dai_hw_params on Built-in Speakers: -2
```

What sticks out to me is that it searches the NHLT for an endpoint on SSP5 but the error from the AVS driver is for SSP0.

**Comment by WeirdTreeThing** [comment]
After some further testing, I figured out the issue was that I had an older version of the topology. Updating my topology to the latest version fixes this issue.

**Comment by xinato** [comment]
| TLDR: Playback attempted on APL-based device using NHLT for non-APL-based device - SKL/KBL.
  |
  | Guess simply updating the Virtual bus ID would do the trick as the rest of the BLOB does not care about it.

 Mr. Rojewski,
I am wondering if the issue you identified above is being addressed upstream. Is there a way I can track the issue to see when it has become a patch? Is there a way for me to see when the patched version becomes an active pre-release version?

Best regards,
And Thank You.

**Comment by xinato** [comment]
Wow! I tried the most recent release of da7219-tplg.bin and max98357a-tplg.bin that were released three weeks ago. The avs driver is now working very well with my nami board.

**Comment by xinato** [comment]
AVS continues to work smoothly for me with the current tplg files.

**Comment by crojewsk-intel** [comment]
Wonderful : )

> | TLDR: Playback attempted on APL-based device using NHLT for non-APL-based device - SKL/KBL. | | Guess simply updating the Virtual bus ID would do the trick as the rest of the BLOB does not care about it.
> 
> Mr. Rojewski, I am wondering if the issue you identified above is being addressed upstream. Is there a way I can track the issue to see when it has become a patch? Is there a way for me to see when the patched version becomes an active pre-release version?
> 
> Best regards, And Thank You.

The end-user support i.e.: waterfalling topologies, kconfigs and all other collaterals to end-users of Fedora, Debian, Ubuntu and other major distros by default is still before us as one would be surprised how many customers come back to us requesting support for yet another configuration on older systems.

Till that happens, we will be monitoring Issues-tab on all avs-related projects.
